cranelift-codegen/src/regalloc/branch_splitting.rs

```diff
@@ -0,0 +1,228 @@
+//! Split the outgoing edges of conditional branches that pass parameters.
+//!
+//! The `minimal` and `tree` register allocators require this.  They could integrate the splitting
+//! in their operation but it adds complexity to do so, esp once live ranges are involved.
+
+use std::vec::Vec;
+
+use crate::cursor::{Cursor, EncCursor};
+use crate::dominator_tree::DominatorTree;
+use crate::flowgraph::ControlFlowGraph;
+use crate::ir::{Ebb, Function, Inst, InstBuilder, InstructionData, Opcode, Value};
+use crate::isa::TargetIsa;
+use crate::topo_order::TopoOrder;
+
+pub fn run(
+    isa: &TargetIsa,
+    func: &mut Function,
+    cfg: &mut ControlFlowGraph,
+    domtree: &mut DominatorTree,
+    topo: &mut TopoOrder,
+) {
+    let mut ctx = Context {
+        new_blocks: false,
+        cur: EncCursor::new(func, isa),
+        domtree,
+        topo,
+        cfg,
+    };
+    ctx.run()
+}
+
+struct Context<'a> {
+    // True if new blocks were inserted.
+    new_blocks: bool,
+
+    // Current instruction as well as reference to function and ISA.
+    cur: EncCursor<'a>,
+
+    // References to contextual data structures we need.
+    domtree: &'a mut DominatorTree,
+    topo: &'a mut TopoOrder,
+    cfg: &'a mut ControlFlowGraph,
+}
+
+impl<'a> Context<'a> {
+    fn run(&mut self) {
+        // Any ebb order will do.
+        self.topo.reset(self.cur.func.layout.ebbs());
```

Comment:
Unfortunately no, because the `self.cur` is used in a mutable way later on, and it is not possible to have a reference to `func` as the cursor already has a reference to it.

However, we do not need to iterate over the newly generated blocks, so I replaced this code by `self.domtree.cfg_postorder().iter().rev()`, the reverse is not needed but would change the order of the generated blocks at the end.

Reply:
hum … never mind, I thought I could but apparently not, as we do not always compute the `postorder` vector when we compute the dominator tree.  So without cloning the topological order is the only one I know.  Except maybe an integer iteration over the `self.cur.layout.ebbs()`